designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,99 @@
+
+## Non-Goals
+
+* This will not be a 1:1 replacement for any of NuGet's existing `IVs` interfaces.
+
+## Solution
+
+```cs
+public interface INuGetProjectServices
+{
+    /// <summary>Gets the list of packages installed in a project.</summmry>
+    /// <param name="project">Path and filename to project file.</param>
+    /// <param name="options">Options to use. If null, uses GetInstalledPakcagesOptions.Default.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>
+    /// The list of packages in the project.
+    /// If the project does not exist in the solution, throws ArgumentException.
+    /// If the project is not ready, throws NuGetProjectNotReadyException.
+    /// </return>
+    Task<IReadOnlyList<NuGetPackage>> GetInstalledPackagesAsync(string project, GetInstalledPackagesOptions options, CancellationToken cancellationToken);
+}
+
+public sealed class NuGetPackage
+{
+    /// <summary>The package id.</summary>
+    string Id { get; }
+
+    /// <summary>The package version</summary>
+    /// <remarks>
+    /// If the project uses packages.config, this will be the installed package version.
+    /// If the project uses PackageReference, this would ideally be the resolved package version, but may be the requested package version.
+    /// </remarks>
+    string Version { get; }
+
+    internal NuGetPackage(string id, string version);
+}
+
+public sealed class GetInstalledPackagesOptions
+{
+    public static GetInstalledPackageOptions Default { get; }
+
+    // in the future may include options such as WaitForNomination, WaitForRestore.
+}
+
+public sealed class NuGetProjectNotReadyException : InvalidOperationException
+{
+}
+
+```
+
+## Future Work
+
+* APIs for:
+  * InstallPackageAsync
+  * UninstallPackageAsync
+
+## Open Questions
+
+* How to handle errors
+  * CPS projects (SDK-style .NET projects) work by: On project load, CPS sends NuGet a "nomination" with project/package information. Therefore, NuGet can't return packages in GetPackagesAsync until CPS nomination complete. With IVsPackageInstallerServices, we throw InvalidOperationException if we haven't got nomination info. Given that the new API needs to work over remoting scenarios, how should it be done?  There are multiple options
+    1. Throw exception (task in faulted state). This might appear as a RemotingException to callers, so maybe hard to distinguish between different types of exceptions?
+    2. Return additional error information. This might be a tuple or a result class. Should we use a type extending `Exception` (stack trace probably wouldn't transport), or have `string Code` and `string Message`?  Or just a `bool Success`, without details about why it was unsuccessful?
+* GetInstalledPackagesAsync
+  * Should returned package version be the resolved version, or the requested range, or the lower bound of the requested range? Leaning towards requested range. It means 
```

Comment:
Personally, i think we sohuld do the installed version only.
We are not doing that right now for PR, so maybe the first implementation can just do the same thing as before.